### PR TITLE
[FSDP][optim_state_dict] Use all_gather to deal with uneven size tensors

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92992
* __->__ #92991

The current `_all_gather_optim_state` pads the uneven tensors which is not necessary as `all_gather` support the uneven tensors. This PR removes the padding logic.

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks!

**Review by awgu — APPROVED** [review]
